File: LeetCode/Delete_Node_in_BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *findMaximum(TreeNode *root){
        TreeNode *iter = root;
        if(!root)return root;
        else{
            while(iter->left){
                iter = iter->left;
            }
        }
         return iter;       
    }
    void printInorder1(TreeNode *root){
        if(!root) return ;
        else{
            printInorder1(root->left);
            cout<<" "<<root->val;
            printInorder1(root->right);
        }
    }
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(!root)
            return NULL;
        else if(root->val < key){
                root->right = deleteNode(root->right,key);
            }
        else if(root->val > key){
                root->left = deleteNode(root->left,key);
            }
        else if(root->val == key){
            if(!root->left && root->val == key){
                return (root->right);
            }
            else if(!root->right && root->val == key){
                return (root->left);
            }
            else{
                TreeNode * maximumNode = findMaximum(root->right);
                root->val = maximumNode->val;
                root->right = deleteNode(root->right,maximumNode->val);
            }
        }
        return root;
    }
};

File: LeetCode/delete_node_in_BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if(findNode(root,key)){
            root = deleteNodeUtil(root,key);
        }
        return root;
    }
    
    public boolean findNode(TreeNode root,int key){
        if(root == null) return false;
        else{
            if(root.val > key){
                return findNode(root.left,key);
            }else if(root.val < key){
                return findNode(root.right,key);
            }else if(root.val == key){
                return true;
            }
        }
        return false;
    }
    
    public TreeNode deleteNodeUtil(TreeNode root, int key){
        if(root == null) return null;
        else{
            if(root.val > key){
                root.left = deleteNodeUtil(root.left,key);
            }else if(root.val < key){
                root.right = deleteNodeUtil(root.right ,key);
            }else if(root.val == key){
                if(root.left == null && root.right == null) return null;
                else if(root.right == null){
                    return root.left;
                }
                else if(root.left == null){
                    return root.right;
                }
                
                TreeNode minimumNode = findMinimum(root.right);
                root.val = minimumNode.val;
                root.right = deleteNodeUtil(root.right,root.val);
                    
            }
        }
        return root;
    }
    
    public TreeNode findMinimum(TreeNode root){
        TreeNode iter = root;
        while(iter.left != null){
            iter = iter.left;
        }
        return iter;
    }
}